### PR TITLE
Helidon Reactive documentation

### PR DESCRIPTION
**Pull request by danielkec**
related PR oracle/helidon-build-tools/pull/90

Signed-off-by: Daniel Kec <daniel.kec@oracle.com>

### CODE REVIEW

File: docs/sitegen.yaml [hunk 1]
```diff
@@ -172,6 +172,14 @@ backend:
...
                 - includes:
                     - "se/websocket/*.adoc"
+            - title: "Reactive Operators"
+              pathprefix: "/se/operators"
+              glyph:
+                type: "icon"
+                value: "fa-filter"
+              items:
+                - includes:
+                    - "mp/reactive/*.adoc"
```
Comment (danielkec) [inline docs/sitegen.yaml]: @m0mus Not sure about this, didn't want to duplicate same doc, operators are exactly the same for both flavors

File: docs/mp/messaging/02_message.adoc [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+///////////////////////////////////////////////////////////////////////////////
+
+= Message
+:description: Helidon MicroProfile Reactive Messaging
+= :keywords: helidon, messaging, reactive, streams
+
+Since Helidon 2.0
```
Comment (ljamen) [inline docs/mp/messaging/02_message.adoc]: The Reactive Messaging [Message] class can be used to wrap or unwrap data items between methods and connectors. The message wrapping and unwrapping can be performed explicitly by using org.eclipse.microprofile.reactive.messaging.Message#of(T) or implicitly through the messaging core.

File: docs/mp/messaging/01_reactive-messaging.adoc [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+= Reactive Messaging
+:description: Helidon MicroProfile Reactive Messaging
+= :keywords: helidon, messaging, reactive, streams
+
+Since Helidon 2.0
+
+== Reactive Messaging
+
+MicroProfile Reactive Messaging provides means for CDI beans to be able produce, consume or process messages over Reactive Streams.
```
Comment (ljamen) [inline docs/mp/messaging/01_reactive-messaging.adoc]: MicroProfile Reactive Messaging uses CDI beans to produce, consume or process messages over Reactive Streams.

File: docs/mp/messaging/01_reactive-messaging.adoc [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+Such messaging bean is expected to be either in `ApplicationScoped` or `Dependent` scope.
+Messages are managed by methods annotated by `@Incoming` and `@Outgoing`, invocation is always driven by messaging core either at assembly time or for every message coming from the stream.
```
Comment (ljamen) [inline docs/mp/messaging/01_reactive-messaging.adoc]: Messages are managed by methods annotated by `@Incoming` and `@Outgoing' and the invocation is always driven by message core - either at assembly time, or for every message coming from the stream.

File: docs/sitegen.yaml (hunk 1 above)
Comment (m0mus) [reply docs/sitegen.yaml]: @danielkec It can be done using includes.

File: docs/mp/messaging/01_reactive-messaging.adoc (hunk 3 above)
Comment (danielkec) [reply docs/mp/messaging/01_reactive-messaging.adoc]: thx

File: docs/mp/messaging/02_message.adoc (hunk 2 above)
Comment (danielkec) [reply docs/mp/messaging/02_message.adoc]: thx

File: docs/sitegen.yaml (hunk 1 above)
Comment (romain-grecourt) [reply docs/sitegen.yaml]: See https://asciidoctor.org/docs/user-manual/#include-directive

Comment (danielkec) [reply docs/sitegen.yaml]: Aha took me long sorry :man_facepalming:

**Comment by danielkec** [comment]
@m0mus Rebased on top of latest doc changes, added Multi/Single doc

File: docs/sitegen.yaml (hunk 1 above)
Comment (m0mus) [reply docs/sitegen.yaml]: It's more a question than suggestion. Would it be better to put shared pieces in `shared` folder and reference them from `se` and `mp`? In this case it's easy to find what is shared.,

**Comment by m0mus** [comment]
@danielkec @ljamen Repeating my question to both of you. I posted it as one of the comments, but after the last commit it got collapsed and not seen anymore. So, I'm not sure that you've seen it.
Would it be better to put shared pieces in `shared` folder and reference them from `se` and `mp` instead of referencing `se` parts from `mp`? In this case it will be easy to identify what is shared.

**Comment by danielkec** [comment]
> @danielkec @ljamen Repeating my question to both of you. I posted it as one of the comments, but after the last commit it got collapsed and not seen anymore. So, I'm not sure that you've seen it.
> Would it be better to put shared pieces in `shared` folder and reference them from `se` and `mp` instead of referencing `se` parts from `mp`? In this case it will be easy to identify what is shared.

I like `share` folder approach

File: docs/se/reactivestreams/02_engine.adoc [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+|flatMapIterable|Transform item with supplied function and flatten resulting `Iterable` to downstream.
+|observeOn|Re-emit the upstream's signals to the downstream on the given executor's thread using a default buffer size of 32 and errors skipping ahead of items.
+|observeOn|Re-emit the upstream's signals to the downstream on the given executor's thread.
+|flatMapIterable|Transform item with supplied function and flatten resulting `Iterable` to downstream.
+|forEach|Terminal stage, invokes provided consumer for every item in the stream.
+|collectList|Collect the items of this `Multi` instance into a `Single` of `List`.
+|collect|Collect the items of this `Multi` instance into a `Single`.
+|collect|Collect the items of this `Multi` into a collection provided via a `Supplier`
+*and mutated by a {@code BiConsumer} callback.
+|collectStream|Collects up upstream items with the help of a the callbacks of a `java.util.stream.Collector`.
```
Comment (akarnokd) [inline docs/se/reactivestreams/02_engine.adoc]: `help of a the callbacks` -> `help of the callbacks`

File: docs/se/reactivestreams/02_engine.adoc [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+|just|Create a `Multi` instance that publishes the given items to a single subscriber.
+|just|Create a `Multi` instance that publishes the given items to a single subscriber.
+|singleton|Create a `Multi` that emits a pre-existing item and then completes.
+|error|Create a `Multi` instance that reports the given exception to its subscriber(s). The exception is reported by invoking `Subscriber#onError(java.lang.Throwable)` when `Publisher#subscribe(Subscriber)` is called.
+|empty|Get a `Multi` instance that completes immediately.
+|never|Get a `Multi` instance that never completes.
+|concat|Concat streams to one.
+|onTerminate|Executes given `java.lang.Runnable` when any of signals onComplete, onCancel or onError is received.
+|onComplete|Executes given `java.lang.Runnable` when onComplete signal is received.
+|onError|Executes given `java.lang.Runnable` when onError signal is received.
```
Comment (akarnokd) [inline docs/se/reactivestreams/02_engine.adoc]: Executes the given `java.util.function.Consumer` when an `onError` signal is received.

File: docs/se/reactivestreams/02_engine.adoc [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+|defaultIfEmpty|Signals the default item if the upstream is empty.
+|switchIfEmpty|Switch to the other publisher if the upstream is empty.
+|peek|Invoke provided consumer for every item in stream.
+|distinct|Filter out all duplicates.
+|filter|Filter stream items with provided predicate.
+|takeWhile|Take the longest prefix of elements from this stream that satisfy the given predicate. As long as predicate returns true, items from upstream are sent to downstream, when predicate returns false stream is completed.
+|dropWhile|Drop the longest prefix of elements from this stream that satisfy the given predicate. As long as predicate returns true, items from upstream are NOT sent to downstream but being dropped, predicate is never called again after it returns false for the first time.
+|limit|Limit stream to allow only specified number of items to pass.
+|skip|Skip first n items, all the others are emitted.
+|flatMap|Transform item with supplied function and flatten resulting `Flow.Publisher` to downstream.
```
Comment (akarnokd) [inline docs/se/reactivestreams/02_engine.adoc]: Transform each upstream item with the supplied function into a `Flow.Publisher`, subscribe to them and then flatten their items into a single sequence of items emitted to the downstream.

File: docs/se/reactivestreams/02_engine.adoc [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+|flatMap|Transform item with supplied function and flatten resulting `Flow.Publisher` to downstream while limiting the maximum number of concurrent inner `Flow.Publisher`s and their in-flight item count, optionally aggregating and delaying all errors until all sources terminate.
```
Comment (akarnokd) [inline docs/se/reactivestreams/02_engine.adoc]: Transform each upstream item with the supplied function and flatten the resulting `Flow.Publisher` [...]

File: docs/se/reactivestreams/02_engine.adoc [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+|flatMapIterable|Transform item with supplied function and flatten resulting `Iterable` to downstream.
```
Comment (akarnokd) [inline docs/se/reactivestreams/02_engine.adoc]: Transform each upstream item with the supplied function and flatten the resulting `Iterable` to the downstream.

File: docs/se/reactivestreams/02_engine.adoc [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+|observeOn|Re-emit the upstream's signals to the downstream on the given executor's thread using a default buffer size of 32 and errors skipping ahead of items.
+|observeOn|Re-emit the upstream's signals to the downstream on the given executor's thread.
+|flatMapIterable|Transform item with supplied function and flatten resulting `Iterable` to downstream.
```
Comment (akarnokd) [inline docs/se/reactivestreams/02_engine.adoc]: Transform each upstream item with the supplied function and flatten the resulting `Iterable` to the downstream.

File: docs/se/reactivestreams/02_engine.adoc [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+|onError|Executes given `java.lang.Runnable` when onError signal is received.
+|onCancel|Executes given `java.lang.Runnable` when a cancel signal is received.
+|takeUntil|Relay upstream items until the other source signals an item or completes.
+|range|Emits a range of ever increasing integers.
+|rangeLong|Emits a range of ever increasing longs.
+|timer|Signal 0L and complete the sequence after the given time elapsed.
+|interval|Signal 0L, 1L and so on periodically to the downstream.
+|interval|Signal 0L after an initial delay, then 1L, 2L and so on periodically to the downstream.
+|timeout|Signals a `TimeoutException` if the upstream doesn't signal the next item, error or completion within the specified time.
+|timeout|Switches to a fallback single if the upstream doesn't signal the next item, error or completion within the specified time.
```
Comment (akarnokd) [inline docs/se/reactivestreams/02_engine.adoc]: Switches to a fallback source if [...]

File: docs/se/reactivestreams/02_engine.adoc (hunk 5 above)
Comment (danielkec) [reply docs/se/reactivestreams/02_engine.adoc]: thx

File: docs/se/reactivestreams/02_engine.adoc (hunk 6 above)
Comment (danielkec) [reply docs/se/reactivestreams/02_engine.adoc]: thx

File: docs/se/reactivestreams/02_engine.adoc (hunk 7 above)
Comment (danielkec) [reply docs/se/reactivestreams/02_engine.adoc]: thx

File: docs/se/reactivestreams/02_engine.adoc (hunk 8 above)
Comment (danielkec) [reply docs/se/reactivestreams/02_engine.adoc]: thx

File: docs/se/reactivestreams/02_engine.adoc (hunk 9 above)
Comment (danielkec) [reply docs/se/reactivestreams/02_engine.adoc]: thx

File: docs/se/reactivestreams/02_engine.adoc (hunk 10 above)
Comment (danielkec) [reply docs/se/reactivestreams/02_engine.adoc]: thx

File: docs/se/reactivestreams/02_engine.adoc (hunk 11 above)
Comment (danielkec) [reply docs/se/reactivestreams/02_engine.adoc]: thx

File: docs/se/reactivestreams/01_introduction.adoc [hunk 12]
```diff
@@ -16,10 +16,33 @@
 
 ///////////////////////////////////////////////////////////////////////////////
 
-= Reactive Streams Operators
+= Overview
 :toc:
 :toc-placement: preamble
-:description: Reactive Streams Operators support in Helidon SE
+:parentTitle: Se & Mp
```
Comment (danielkec) [inline docs/se/reactivestreams/01_introduction.adoc:22]: @m0mus Parent title override will kick in when oracle/helidon-build-tools/pull/92 is used for site creation

**Comment by danielkec** [comment]
@akarnokd @ljamen Lists of operators are little stark now, I am planning to expand operators documentation in future PR's with fancy marble diagrams. Would like to just lay the ground structure in this PR so it wouldn't grow humongous and unreviewable.

**Comment by danielkec** [comment]
Continuation of https://github.com/oracle/helidon-build-tools/pull/92#issuecomment-607388688
@romain-grecourt 
> The title in your screenshot is "Se & Mp". I thought you are now using distinct documents and including shared content using the includes directive.
>
Yes I do, thats the `parentTitle` override oracle/helidon-build-tools/pull/92 is about
> If they are separate document, there is one for se and one for mp right ? If so then why is the title "Se & Mp" ?
> 
I am forced to use separate documents by technical reasons, but its functionality available in both flawors of Helidon, so using `parentTitle` to let user know.
> Also, the route is 01_introduction but the 2nd part of the title is "Overview", that's not consistent.
>
My bad should be `01_overview`

File: docs/shared/reactivestreams/03_rsoperators.adoc [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+[[terms]]
+.Operators(Stages)
+|===
+|from | Create new PublisherBuilder supplied Publisher
+|of | Create new PublisherBuilder emitting supplied elements
+|ofNullable | Empty stream if supplied item is null
+|iterate | Create infinite stream
+|generate| Create infinite stream
+|empty| Create empty stream
+|failed| Create stream which is send onError as a first thing is
```
[the remark this answers is not in the document]
Comment (danielkec) [reply docs/shared/reactivestreams/03_rsoperators.adoc:51]: yay thanks

File: docs/mp/reactivestreams/01_introduction.adoc [hunk 14]
```diff
@@ -16,11 +16,34 @@
...
+.MicroProfile Reactive Streams Operators
+[icon=fa-book,link=mp/reactivestreams/03_rsoperators.adoc]
+--
+Microprofile implementation.
+--
+====
+
+There are two handy apis for working with reactive streams available in {helidon},
+one for working with `java.util.concurrent.Flow`
+and second for `org.reactivestreams` based reactive components.
```
[the remark this answers is not in the document]
Comment (danielkec) [reply docs/mp/reactivestreams/01_introduction.adoc]: thx

File: docs/shared/reactivestreams/02_engine.adoc [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+///////////////////////////////////////////////////////////////////////////////
+= Reactive Engine
+== Reactive Engine
+
+Helidon has its own dependency-less(no dependencies outside of Helidon ecosystem) set of reactive operators for work with `java.util.concurrent.Flow` based reactive streams.
```
[the remark this answers is not in the document]
Comment (danielkec) [reply docs/shared/reactivestreams/02_engine.adoc]: thx

**Comment by paulparkinson** [comment]
We still need to add the SE messaging examples or I am missing them? Otherwise looks good so far.

**Pull request merged**